src/maxdiffusion/models/attention_flax.py

```diff
@@ -18,13 +18,237 @@
 import flax.linen as nn
 import jax
 import jax.numpy as jnp
+from jax.experimental import shard_map
+from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_mask
+from jax.experimental.pallas.ops.tpu.splash_attention import splash_attention_kernel
 
 from maxdiffusion import common_types
 
+Array = common_types.Array
+Mesh = common_types.Mesh
+DType = common_types.DType
+
+
 AxisNames = common_types.AxisNames
 BATCH = common_types.BATCH
 LENGTH = common_types.LENGTH
 HEAD = common_types.HEAD
+D_KV = common_types.D_KV
+
+class AttentionOp(nn.Module):
+    mesh: Mesh
+    attention_kernel: str
+    scale: int
+    heads: int
+    dim_head: int
+    use_memory_efficient_attention: bool = False
+    split_head_dim: bool = False
+    float32_qk_product: bool = True
+    flash_axis_names: AxisNames = (BATCH, HEAD, LENGTH, D_KV)
+    flash_min_seq_length: int = 4096
+    dtype: DType = jnp.float32
+
+    def check_attention_inputs(
+        self,
+        query: Array,
+        key: Array,
+        value: Array
+    ) -> None:
+
+        """Check attention inputs."""
+
+        assert key.ndim == value.ndim, 'k, v must have same rank.'
+        assert query.shape[:-3] == key.shape[:-3] == value.shape[:-3], (
+            'q, k, v batch dims must match.')
+        assert key.shape[-2] == value.shape[-2], ('k, v num_kv_heads must match.')
+        assert key.shape[-3] == value.shape[-3], 'k, v lengths must match.'
+        assert query.shape[-1] == key.shape[-1], 'q, k depths must match.'
+
+    def apply_attention(
+        self,
+        query: Array,
+        key: Array,
+        value: Array,
+        decoder_segment_ids: Array | None = None
+    ):
+        self.check_attention_inputs(query, key, value)
+
+        can_use_flash_attention = query.shape[1] >= self.flash_min_seq_length and key.shape[1] >= self.flash_min_seq_length and value.shape[1] >= self.flash_min_seq_length
+
+        if self.attention_kernel == "dot_product" or self.use_memory_efficient_attention or not can_use_flash_attention:
+            return self.apply_attention_dot(query, key, value)
+        elif self.attention_kernel == "flash":
+            return self.tpu_flash_attention(query, key * self.scale, value, decoder_segment_ids)
+        else:
+            raise ValueError(f'Unexpected attention kernel {self.attention_kernel=}.')
+
+    def tpu_flash_attention(
```

Comment:
can we make sure that every function has comments? and also adding comments for the stuff we discussed so that we don't forget why we added some of the parts we added?